### ISSUE TITLE
Feature request: Ignored branches for deploy previews

### ISSUE THREAD

**Issue by Betree** [vercel/vercel#3166]
Hi there!

Zeit deploy previews are awesome, it's definitely an essential feature for our workflow at Open Collective.

We're using a tool called Crowdin to manage our translations. This tool creates a lot of commits, sometimes hundreds in the same day (see https://github.com/opencollective/opencollective-frontend/pull/2824). Deploys preview don't really matter for the pull requests created by this tool and we would prefer to not flood our Zeit pre-deploys with it.

**Proposed solution**

In `now.json`, add a `deploy-previews` fields that would accept an `exclude` parameter like this:

```json
{
  "version": 1,
  "name": "opencollective-frontend",
  "type": "docker",
  "deploy-previews": {
    "exclude": ["i18n/crowdin"]
  }
}
```

Now would then not deploy any branch matching the names given in this array.

Bonus: allow wildcards in names (ex: `i18n/*`) - not required in our use case, but could be nice.

**Comment by Betree** [comment]
As an alternative, an `exclude-labels` setting with a list of labels that will prevent Zeit from deploying would do the job just as fine

**Comment by styfle** [comment]
Deploys happen on push, not on commit.

Is it possible to adjust your tool to only push once for the PR?

**Comment by Betree** [comment]
We can indeed change the push interval to reduce the frequency (1 per day for example) but that's not very convenient for us as we often want do last-minute translations before the releases so we like that it's pushing things fast (currently up to 1push/10minutes).

**Comment by styfle** [comment]
It looks like you are still using Now 1.0 which is typically slower than Now 2.0 because it has to build a complete docker image.

Have you considered [upgrading to 2.0](https://zeit.co/docs/v2/advanced/platform/changes-in-now-2-0)?

**Comment by leo** [comment]
Thank you for creating this issue!

Please note that – in order for us to be able to handle each feature request with care – we have to consume all of them through a unified pipeline that makes it easier for us to prioritize, track and progress on the features our user base is interested in.

In turn, the [Issues tab](https://github.com/zeit/now/issues) on this repository is best only be used for reporting bugs, which we can immediately act on.

For well-explained feature requests (like the one you just posted), please contact us at [support@zeit.co](mailto:support@zeit.co), so that we can consider adding it to our roadmap.

**Comment by Betree** [comment]
> It looks like you are still using Now 1.0 which is typically slower than Now 2.0 because it has to build a complete docker image.
> 
> Have you considered [upgrading to 2.0](https://zeit.co/docs/v2/advanced/platform/changes-in-now-2-0)?

Sorry for the late reply! I'm not the one working on it, but it seems that we can't update to 2.0 because we're using `express`.

> Thank you for creating this issue!
> 
> Please note that – in order for us to be able to handle each feature request with care – we have to consume all of them through a unified pipeline that makes it easier for us to prioritize, track and progress on the features our user base is interested in.
> 
> In turn, the [Issues tab](https://github.com/zeit/now/issues) on this repository is best only be used for reporting bugs, which we can immediately act on.
> 
> For well-explained feature requests (like the one you just posted), please contact us at [support@zeit.co](mailto:support@zeit.co), so that we can consider adding it to our roadmap.

Before posting this issue I also sent it on the "Feedback" box on Zeit but didn't received any reply.

![image](https://user-images.githubusercontent.com/1556356/72333946-e875b080-36bc-11ea-9e80-8fca3301038f.png)

I think having a public place to show feature requests is great, because it allows people to +1 rather than writing a full feature request for something that was already asked. People can subscribe, and by replying on it everyone's gets the info about whether or not this will be done/prioritized.

That being said, it's understandable that you may want to prefer tracking everything on a single (private) tool to organize better.

**Comment by Andre601** [comment]
Is this a thing now, or still planned to perhaps be implemented, as I myself would like to have this, to prevent ZEIT now from deploying changes when pushed to the gh-pages branch (Which may cause errors apparently)

**Comment by njbarrett** [comment]
I'm still interested in this feature, has anything happened with it?

**Comment by ay8s** [comment]
Would love this also.

**Comment by mylesthedev** [comment]
This would be a great feature. I vote for whitelist approach.

**Comment by clement-faure** [comment]
Same here, would love this.

**Comment by styfle** [comment]
This feature is available today, called [Ignored Build Step](https://vercel.com/docs/platform/projects#ignored-build-step). You can write a custom script to ignore or allow builds after you `git push`, for example ignore a certain branch.

![image](https://user-images.githubusercontent.com/229881/93212181-3f41f700-f730-11ea-8057-cb66f46c7f9e.png)

**Comment by Andre601** [comment]
> This feature is available today, called [Ignored Build Step](https://vercel.com/docs/platform/projects#ignored-build-step). You can write a custom script to ignore or allow builds after you `git push`, for example ignore a certain branch.
> 
> ![image](https://user-images.githubusercontent.com/229881/93212181-3f41f700-f730-11ea-8057-cb66f46c7f9e.png)
> 

As much as this is an option does it require some git knowledge which not everyone heee might offer (also, why am I still subscribed to this? I don't use this anymore)

**Comment by MatyiFKBT** [comment]
> 
> This feature is available today, called [Ignored Build Step](https://vercel.com/docs/platform/projects#ignored-build-step). You can write a custom script to ignore or allow builds after you `git push`, for example ignore a certain branch.
> 
> ![image](https://user-images.githubusercontent.com/229881/93212181-3f41f700-f730-11ea-8057-cb66f46c7f9e.png)

How would you exclude a branch with this feature? 

@styfle  Could you please show an example?

**Comment by njbarrett** [comment]
An example is required here please @styfle

**Comment by alpapp** [comment]
Yes please, an example.  I want to prevent preview builds of the "development" branch of my project.  What command must I use in this Ignored Build Step facility?

Another use case might be preventing all builds except for the master branch.

Any examples welcome!

**Comment by magic-akari** [comment]
I tried the following command, but it doesn't work.

```bash
[[ $(git branch --show-current) = "gh-pages" || $(git diff --quiet HEAD^ HEAD ./) ]]
```

Maybe the command need to be wrapped with `bash -c`.

**Comment by njbarrett** [comment]
I have tried similar commands to the above, but turns out that `git branch --show-current` is always returning `master` even when the deployment is for another branch.

I found this out by creating a shell script that runs before the build and echoing the value out.

**Comment by alpapp** [comment]
Searching google, the following command returns the most recently committed ("pushed" / "synced" ?) branch:
`git for-each-ref --count=1 --sort=-committerdate refs/heads/ --format='%(refname:short)'`

How do we structure this such that the Ignored Build Step function can return a 1 (for make a build) or 0 (ignore it) when it = "master" (or some other branch of your choosing)?

**Comment by alpapp** [comment]
I knocked up a quick nodejs script to exec this command and, as @njbarrett has found, Vercel always returns "master" as the local branch even though the beginning of the build log shows it's coming from another branch.  It would seem the local build environment always places the pulled down code into it's own master branch and that is that.

Vercel team - how can we determine reliably which remote branch the Vercel build process is building for, in code, within the Ignored Build Step command?  As previously established, git branch --show-current and similar all return "master" no matter the originating branch.  (Querying the remote is only a partial solution as it may be a private repository and unwilling to expose login information within IB script)

**Comment by alpapp** [comment]
Further research indicates that the System Environment Variable `VERCEL_GITHUB_COMMIT_REF` should contain the branch name that triggered the build process, accessible via process.env.VERCEL_GITHUB_COMMIT_REF provided you've added it in the project Settings page under Enviroment Variables.  However, I can't get it read anything if used within the Ignored Build Step script (but does read OK within the normal build step that subsequently ensues).  It would seem to me that the env variable is not set for the Ignored Build Step before the actual build is performed, and hence is not useable for this purpose.
[System Environment Variables doc page refers.](https://vercel.com/docs/build-step#system-environment-variables)

**Comment by njbarrett** [comment]
@alpapp I came to similar conclusions also. So my work around for now is to check that env variable in the build step and fail the build if that env variable equals one I dont want it to deploy. So I haven't "ignored" the branches but they will error.

Hopefully someone from Vercel can chime in as to a proper solution using the feature they claim as the solution.

**Comment by styfle** [comment]
Thanks, we'll get this fixed and document a proper solution!

**Comment by michaelbirchler** [comment]
related discussion here https://github.com/vercel/vercel/discussions/5171

**Comment by MatyiFKBT** [comment]
They published this 2 days ago: https://vercel.com/knowledge/how-do-i-use-the-ignored-build-step-field-on-vercel

**Comment by alpapp** [comment]
Yes indeed, except their example script doesn't work as it implies it should.  Any commit to any branch triggers my build, which is clearly not how it's supposed to be.

**Comment by paulogdm** [comment]
> Yes indeed, except their example script doesn't work as it implies it should. Any commit to any branch triggers my build, which is clearly not how it's supposed to be.

Can you elaborate? I've tested the example again and it worked as expected.

**Comment by deakster** [comment]
Personally I haven't tried yet if this works or not, but the implementation sounds bizarre to me.

If the command returns what is conventionally the success exit code (0) the build will halt, whereas if the command returns what is normally an error code (1), the build goes ahead?

Anywhere else, a command that fails with a non-zero exit code would stop the build.

**Comment by paulogdm** [comment]
@deakster Think about the command in its most common use-case:

If `git diff (...)` returns something, the build must continue. If not, you halt it.

**Comment by deakster** [comment]
@paulogdm Ok, not a big deal to me although the command we are running to determine build progress is not "git diff" in our case.

I think @alpapp is correct that it doesn't work though. The blog post [here](https://vercel.com/knowledge/how-do-i-use-the-ignored-build-step-field-on-vercel) suggests that the BRANCH environment variable is available and can be used, but this does not seem to be the case for us. We are running a very simple node script like `node buildcheck.js`, and when we print out the entire contents of `process.env` there is nothing there indicating the BRANCH.

**Comment by paulogdm** [comment]
@deakster Please refer to https://vercel.com/docs/environment-variables#references-to-system-environment-variables. You need to include the variable explicitly. You can use `BRANCH` as a reference to `VERCEL_GIT_COMMIT_REF`.

**Comment by preeteshjain** [comment]
I wish this was as simple as just picking up the branches you want deployments for, in the project settings.

**Comment by realmichaelye** [comment]
+1 for @preeteshjain

**Comment by Tbhesswebber** [comment]
Can we maybe get some clarity on what "ignore" means? It sounds like it should be "don't even try to build", but the implementation seems to be "reject the build"

**Comment by Js-Brecht** [comment]
We need to be able to either turn off deployments for branches, or (preferably) let us list the branches that _should be_ deployed.  As in, do not even start a deployment from a PR/Commit that's not targeting one of the branches configured _for_ deployment.  Just like Netlify does!

I am running a monorepo with > 20 projects, only one of which is deployed on Vercel.  With the Git integration turned on, we get too much noise on PRs that don't even touch Vercel (even when `github.silent: true`).  Yes, I can tell Vercel to _not build_, but that's after the deployment has already been created.  Pretty wild that there's no control over this.

Honestly, Vercel does so many things right, but... it seems like a page or two could be taken from Netlify's book when it comes to git deployments.

I do like the `Ignored Build Step`, though... except that it requires some ridiculous acrobatics when your last commit wasn't one that changes a project being deployed to Vercel, but there is a commit in what you just merged that does.

**Comment by tbgse** [comment]
After reading through all this, I keep wondering if anyone here managed to make the ignore build step feature work in a straightforward way to exclude specific branch patterns? It'd be such a nice feature to have.

**Comment by diecknet** [comment]
@tbgse I use this code as an **"Ignored Build Step"** to exclude one branch named "export" from building with Vercel.
Prerequisite is to enable "Automatically expose System Environment Variables" (under "Environment Variables").
```
bash -c 'if [[ $VERCEL_GIT_COMMIT_REF != "export" ]]; then exit 1; fi'
```

You can also create bash script to exclude more branch names. I'm not so familar with bash, so I don't know how to exclude specific branch patterns, but it's probably not that complicated.
Start with this:
```
#!/bin/bash
echo "VERCEL_GIT_COMMIT_REF: $VERCEL_GIT_COMMIT_REF"
if [[ $VERCEL_GIT_COMMIT_REF != "export" ]]; then 
  exit 1; 
fi
exit 0;
```
And then enter ```bash yourscriptname.sh``` as the "Ignored Build step"

EDIT: Eventhough this workaround prevents the build, I would also love to have a native option to include/exclude build for specific branches or branch patterns.

**Comment by magland** [comment]
I confirmed that the bash command of @diecknet works, and I updated it to only deploy on the "deploy" or "preview" branches with the following command:

```
bash -c 'if [[ $VERCEL_GIT_COMMIT_REF == "deploy" ]] || [[ $VERCEL_GIT_COMMIT_REF == "preview" ]]; then exit 1; fi'
```

Along with others, I am very surprised that the default behavior is to deploy everything! You would think that both vercel and user would benefit from having select targeted deploys be the default.

**Comment by cyrus-za** [comment]
I want all branches to deploy but only affected apps, and its not as simple as checking if a certain directory changes, seeing as the monorepo has packages/libs and some libs consume other libs that are consumed by apps. We use [nx workspaces](https://github.com/nrwl/nx) and they got a great command `nx affected:apps` which spits out the apps that are affected. 

The first problem is that in order to run the command for nx, I need to install all dependencies, which is then part of a build. The second problem is that nx uses git under the hood to check diffs between the current HEAD and master. The problem is that when one does `git branch` it shows that we're on master regardless of which github branch is actually being deployed. 

I know the nx and vercel team have been working closely the past few months, so I just hope that somne native solution gets released soon. Until then we are building every single app on every single commit, whether that app got affected or not and we got over 20 projects in this monorepo. In fact, just yesterday we got a email from Vercel support team that we've hit an unexpected storage limit with AWS (vercel upstream provider) and we need to delete some of our older deploys in order to have more deploys getting built.

If this skip step worked, we would have a fraction of the amount of deploys, so please prioritise this

**Comment by ccrvlh** [comment]
Any news on this? All of `feature` branches are being deploy in Preview, it seems that this is a waste of resources actually (build time), both for us and for Vercel. As someone mentioned, could be as simple as choosing a production branch, and preview branch. Doing it the other way around, "Ignoring", would be complex and not friendly since it would require some `regex` since new branchs with different names and name patterns are create frequently.

 I would be more than happy to have production running with `master` and preview running with `develop` for example, simple as that. There could even be a possibility to make more complex rules for production/deployment available only on more expensive plans.

**Comment by preeteshjain** [comment]
This is a no-brainer, must have feature that saves time, money and resources for both Vercel and users.

Please implement this ASAP.

**Comment by lorenzodejong** [comment]
+1, the proposed solution from @magland works fine overall. However i've noticed that (when using the Github integration) the Github Deployments API is called even though the deployment is canceled from the "Ignore build step" setting. This is especially annoying when using integrations like Github for Slack, because a "Vercel deployment" Slack notification is still triggered even though the deployment is canceled right after starting the deployment run.

I would really like to see first-class support for limiting deployments to specific branches.

**Comment by ImOnMars** [comment]
It's confusing how some of the simplest features that anybody would expect to exist just don't sometimes. Surely adding a couple of toggles and an optional whitelist of branches would be rather straightforward and distinctly essential for a platform that emphasizes developer-focused ease of use and accessibility.

**Comment by s-en-o** [comment]
> 
> 
> It's confusing how some of the simplest features that anybody would expect to exist just don't sometimes. Surely adding a couple of toggles and an optional whitelist of branches would be rather straightforward and distinctly essential for a platform that emphasizes developer-focused ease of use and accessibility.

I think It can be an option on what branch we want to deploy.
i.e:

- Want to deploy `f-staging` branch to a staging domain
- Want to deploy `f-someFeature` branch to a feature domain
- And of course your production branch from `main` or `master` will deploy to primary domain

Other than what branches we have already added to the list do not deploy.

**Comment by elnelsonperez** [comment]
It is surprising to me how this issue is still open after three years.
Is it that hard to implement?

**Comment by preeteshjain** [comment]
@rauchg @leerob @timneutkens It would be great if you guys give attention to this issue, that a lot of people want implemented.

**Comment by dgurns** [comment]
Completely agree with the comments so far. I have `silent: true` implemented, as well as `Ignore Build Step`, but I still see the bot deploy messages on my PR -- even though nothing is being deployed.

<img width="888" alt="Screen Shot 2021-12-31 at 5 07 12 PM" src="https://user-images.githubusercontent.com/1173791/147833519-910e7f2f-ec1c-4682-9c03-c8e50dba2b6a.png">

I'd propose that if a build has been skipped, then the Vercel bot **should not** post a deploy message on the PR.

Any chance of implementing that, Vercel team? (Love the product otherwise)

Outside of that change, anyone been able to figure out how to silence this?

**Comment by KaneLabs** [comment]
I understand dev time is a scarce resource but now we can't use vercel for our monorepo because of these limitations.

**Comment by msw93** [comment]
Any updates on this feature? A potential timeline for resolution?

**Comment by b-lov** [comment]
only missing feature, that makes me use netlify over vercel

**Comment by stvnbash** [comment]
It would be a great feature to be able to select specific branches for Preview Deployments similar to how we already can select a branch for Production Deployments.  I commit/push relatively frequently as I'm developing--and I don't need every push on development branches creating a new Preview Deployment.

**Comment by lucavenir** [comment]
I'm honestly surprised this is not a thing as today. It just makes sense to select which branches I need to build and spend resources on, as most of the times as a developer I just work on feature or local development branches. Just because I'm pushing some changes doesn't mean that I need a live preview of a dev branch. Or at least this should be customizable.

**Comment by rayjasson98** [comment]
Another problem is that any failed/ignored build step will appear as a "Build Cancelled" deployment in the build logs. This makes it quite hard to find out which deployment succeeds or fails, as you will need to rummage through all those "Build Cancelled" logs.

**Comment by tobiastornros** [comment]
Would be nice to have deployments only for open PRs, staging and production.

**Comment by ZetiMente** [comment]
Yes, was using Netlify but now rethinking things. This feature was simple UI configuration over there. But here is quite confusing.

**Comment by hormesiel** [comment]
Here's a simplified version of the command in the previous comments to use in `Ignored Build Step` :

```bash
[ $VERCEL_GIT_COMMIT_REF != "preview" ] && [ $VERCEL_GIT_COMMIT_REF != "production" ]
```

This will trigger (i.e. not cancel) new deployments for the `preview` and `production` branches only.

**Comment by ghost** [comment]
With Firebase backend, it is necessary to specify the domain to use App Check and recaptcha, so it is stressful that it will be deployed every time you commit and an authentication error will occur, so please solve it as soon as possible.

**Comment by liblaf** [comment]
Maybe the [`git.deploymentEnabled`](https://vercel.com/docs/project-configuration#git-configuration/git-deployment-enabled) option in [`vercel.json`](https://vercel.com/docs/project-configuration#git-configuration/git-deployment-enabled) would do?

**Comment by oliversturm** [comment]
Potential paying customer reading here... off to try the suggested workarounds now, but ... seriously? Doesn't look promising for your support that a simple and obviously required feature like this is not implemented after 3.5 years.

**Comment by vincentjflorio** [comment]
I would also like this; I appreciate that I don't know the magic behind the scenes that it might take, but in my layperson brain it just seems like it'd be a covered base.

"Here is a thing, and here is what you can do it to."

For various other non-Vercel reasons I need to update my repo, pushing things from local to GitHub so I can't stop doing that.

If Vercel isn't in a position to stop from building each one it sees, is there perhaps a means on GitHub of hiding from Vercel on a per-branch basis? :)

**Comment by qrkourier** [comment]
This has been a thorn in my side despite the availability of `ignoreCommand` in vercel.json and "Ignored Build Command" in Git settings.

The problem is that `vercel.json` isn't available on the branch I want to ignore, so I must use project settings, not vercel.json to configure the ignore command, and the "Ignored Build Command" BASH test doesn't seem to have access to the branch name in `VERCEL_GIT_COMMIT_REF`, despite having opted-in to "Automatically expose System Environment Variables" in "Environment Variables" settings.

My ignored build command in "Git" "Project Settings":

```bash
[[ "${VERCEL_GIT_COMMIT_REF}" =~ ^(gh-pages)$ ]]
```

**Comment by alexander-mart** [comment]
Answer in docs:
https://vercel.com/guides/how-do-i-use-the-ignored-build-step-field-on-vercel#with-a-script

@leo this issue can be closed? :)

**Comment by preeteshjain** [comment]
@alexander-mart No, this cannot be closed. This is not the ideal solution if you have read the comments above. This is feature request, and Vercel hasn't addressed this yet. Read https://github.com/vercel/vercel/issues/3166#issuecomment-882289389

**Comment by yannispgs** [comment]
Vercel team does not seem to take Feature Requests from this Repository Issues. Maybe the creator of the issue (@Betree) or someone else in this topic could create such post [in the dedicated section](https://github.com/vercel/community/discussions/categories/ideas).

If the newly created post managed to have lots of likes/thumbsups, maybe the development team will pay more attention to the matter.

I already opened [a case](https://github.com/vercel/community/discussions/1715) that could match with some of the above use cases for those who seek a more powerful `git.deploymentEnabled` feature.

**Comment by ShanArosh** [comment]
It seems we would have to rely on scripts for a long time.
If anyone is using Turborepo and wants only to allow certain branches to be built, feel free to use this script as `Ignore Build Command`.
## build command
> bash <path-to-script.sh>
```
bash script.sh
```
## script.sh
 
```
#!/bin/bash

echo "VERCEL_GIT_COMMIT_REF: $VERCEL_GIT_COMMIT_REF"

if [[ "$VERCEL_GIT_COMMIT_REF" == "main" || "$VERCEL_GIT_COMMIT_REF" == "sandbox"  ]] ; then
  if npx turbo-ignore 2> /dev/null; then
    # Don't build
    echo "🛑 - Build cancelled"
    exit 0;
  else
    # Proceed with the build
    echo "✅ - Build can proceed"
    exit 1;
  fi
else
  # Don't build
  echo "🛑 - Build cancelled"
  exit 0;
fi


# https://vercel.com/guides/how-do-i-use-the-ignored-build-step-field-on-vercel
```
This script will only build if the branch is `main` or `sandbox` and `npx turbo-ignore` leverages the Turborepo dependency graph to automatically determine if each app, or one of its dependencies has changed and needs to be deployed.

**Comment by walsha2** [comment]
> The problem is that `vercel.json` isn't available on the branch I want to ignore, so I must use project settings, not vercel.json to configure the ignore command

@qrkourier This is 100% the issue and why @preeteshjain says this needs to be a **feature**. The current `Ignored Build Command` logic is way too limiting and does not/cannot handle every case.

The `gh-pages` issue is particularly annoying. The `gh-pages` branch does not have any source code typically, so there is no `vercel.json` or bash script that can be executed. Further, the `Root Directory` that Vercel expects likely does not exist on that branch. Attempts at setting the value in the Vercel project settings does not work either. No matter how many times I try, this never works because it errors out saying the expected deployment location cant be found. Its almost like it checks `Root Directory` setting **BEFORE** it checks `Ignored Build Step` which is odd.

```
echo $VERCEL_GIT_COMMIT_REF; if [ "$VERCEL_GIT_COMMIT_REF" == "gh-pages" ]; then exit 0; else exit 1; fi
```

A nice solution for the community is to simply allow Project level settings to explicitly ignore certain branch names, or expand this `Ignored Build Step` logic to more than its simplistic one-line input. Relying on `vercel.json` or a script in the branch to exist **is not** viable for all use cases.

--------

TLDR: The`Ignored Build Step` feature in Vercel needs to be expanded. Please stop referring users with this issue to the existing docs or examples. They do not solve the problem.

**Comment by bezugen** [comment]
any news on this issue? it's one of the most commented issues, it's already 4 years and it's still happening :-(

**Comment by nextor2k** [comment]
It's almost 2024, and this feature still needs to be corrected!
